### ISSUE TITLE
Warn if code and schema are of different versions

### ISSUE THREAD

**Issue by jrha** [quattor/ncm-ncd#41]
Modules using data from the configuration tree should issue a warning if the version of their Perl module is different from the schema version used to validate their tree, this can occur if sites do not pin their package versions or build modified packages.

This is a high level tracker to track the implementation of this across repositories as required.

**Comment by jrha** [comment]
Many components already provide a version number such as:
```pan
"/software/components/pam/version" = '${no-snapshot-version}';
```
Which might help in the implementation of this.

**Comment by ned21** [comment]
Please make this optional since our package upgrade schedules are not synchronised to our template releases.

**Comment by stdweird** [comment]
@ned21 can you clarify a bit more? are you running schemas that are out of sync with components? who puts the schema in place (maunal or scripted?). and how do you expect the component to do the right thing?

decoupling schema and component is not for the faint of heart, we had our share of self inflicted issues and that's the main reason i'd like to have this.

regardless, ofcourse this will be an opt-in possibility to prevent you from doing wrong things. i see this in a first phase only logging warning or error, not anything more than that.

**Comment by ned21** [comment]
Running schemas that are backwards and forwards compatible with different code releases is quite common in the database world - it's how people do A/B testing, etc.  We just take care to only deploy breaking schema changes when we know a code change has propagated to our plant.

We're managing ok so far but it does take us ages to upgrade and we are currently doing it component by component since we were moving from the pre-unit test versions versions to the maven built code base.

**Comment by jrha** [comment]
Indeed! But it makes sense to have a reporting mechanism for the versions (and when they differ) which is what this is about.
